backend/src/schema/resolvers/follow.js

```diff
@@ -1,51 +1,44 @@
+import { neode as getNeode } from '../../bootstrap/neo4j'
+
+const neode = getNeode()
+
 export default {
   Mutation: {
     follow: async (_object, params, context, _resolveInfo) => {
-      const { id, type } = params
-
-      const session = context.driver.session()
-      const transactionRes = await session.run(
-        `MATCH (node {id: $id}), (user:User {id: $userId})
-          WHERE $type IN labels(node) AND NOT $id = $userId
-          MERGE (user)-[relation:FOLLOWS]->(node)
-          RETURN COUNT(relation) > 0 as isFollowed`,
-        {
-          id,
-          type,
-          userId: context.user.id,
-        },
-      )
-
-      const [isFollowed] = transactionRes.records.map(record => {
-        return record.get('isFollowed')
-      })
+      const { id: followedId, type } = params
+      const { user: currentUser } = context
 
-      session.close()
+      if (type === 'User' && currentUser.id === followedId) {
+        return null
+      }
 
-      return isFollowed
+      const [user, followedNode] = await Promise.all([
+        neode.find('User', currentUser.id),
+        neode.find(type, followedId),
+      ])
+      await user.relateTo(followedNode, 'following')
+      return followedNode.toJson()
     },
 
     unfollow: async (_object, params, context, _resolveInfo) => {
-      const { id, type } = params
-      const session = context.driver.session()
+      const { id: followedId, type } = params
+      const { user: currentUser } = context
 
-      const transactionRes = await session.run(
-        `MATCH (user:User {id: $userId})-[relation:FOLLOWS]->(node {id: $id})
+      /*
+       * Note: Neode doesn't provide an easy method for retrieving or removing relationships.
+       * It's suggested to use query builder feature (https://github.com/adam-cowley/neode/issues/67)
+       * However, pure cypher query looks cleaner IMO
+       */
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-18T22:03:54Z" title="Thursday, September 19th 2019, 12:03:54 am +02:00">Sep 19, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,51 +1,44 @@
+import { neode as getNeode } from '../../bootstrap/neo4j'
+
+const neode = getNeode()
+
 export default {
   Mutation: {
     follow: async (_object, params, context, _resolveInfo) => {
-      const { id, type } = params
-
-      const session = context.driver.session()
-      const transactionRes = await session.run(
-        `MATCH (node {id: $id}), (user:User {id: $userId})
-          WHERE $type IN labels(node) AND NOT $id = $userId
-          MERGE (user)-[relation:FOLLOWS]->(node)
-          RETURN COUNT(relation) > 0 as isFollowed`,
-        {
-          id,
-          type,
-          userId: context.user.id,
-        },
-      )
-
-      const [isFollowed] = transactionRes.records.map(record => {
-        return record.get('isFollowed')
-      })
+      const { id: followedId, type } = params
+      const { user: currentUser } = context
 
-      session.close()
+      if (type === 'User' && currentUser.id === followedId) {
+        return null
+      }
 
-      return isFollowed
+      const [user, followedNode] = await Promise.all([
+        neode.find('User', currentUser.id),
+        neode.find(type, followedId),
+      ])
+      await user.relateTo(followedNode, 'following')
+      return followedNode.toJson()
     },
 
     unfollow: async (_object, params, context, _resolveInfo) => {
-      const { id, type } = params
-      const session = context.driver.session()
+      const { id: followedId, type } = params
+      const { user: currentUser } = context
 
-      const transactionRes = await session.run(
-        `MATCH (user:User {id: $userId})-[relation:FOLLOWS]->(node {id: $id})
+      /*
+       * Note: Neode doesn't provide an easy method for retrieving or removing relationships.
+       * It's suggested to use query builder feature (https://github.com/adam-cowley/neode/issues/67)
+       * However, pure cypher query looks cleaner IMO
+       */
```


Agree!

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-18T22:05:19Z" title="Thursday, September 19th 2019, 12:05:19 am +02:00">Sep 19, 2019</time>_

---


I haven't made a final decision about it, ie. if we should use it for production code or if we should only use it for our factories. :thinking: :woman_shrugging: